### ISSUE TITLE
Provenance publication & discovery

### ISSUE THREAD

**Issue by joshuagl** [slsa-framework/slsa#269]
Like Policy & Verification in https://github.com/slsa-framework/slsa/issues/46, and related to, how/where to publish provenance and how users discover published provenance feels like an open gap in the current set of guidance around SLSA.

I think we should capture recommendations/thoughts around how provenance attestations could be published and discovered for different scenarios.

Today folks can:
* publish alongside their release artefacts, which isn't especially discoverable
* store attestations in sigstore's rekor, discovery becomes a rekor query

One possibility is package ecosystems publishing provenance attestations alongside the packages and package managers growing features to fetch the associated attestations.

Has anyone thought about this in detail?

**Comment by mlieberman85** [comment]
Yes, I have given this a ton of thought over the past few months. Let me see what I can write up.

My current two cents is that discovery should for now very much be in the hands of the client. It should be up to them what signatures they trust and what data stores they want to check for attestations in, beyond any attestations that could live in or alongside the package, e.g. OCI repos having attestations live in the repo along with the blobs and images.

**Comment by MarkLodato** [comment]
Agreed on the need for this. My thinking is that we probably want *both* "publish alongside release artifacts" *and* "store attestations in a public database" (be it rekor or something else). The former is valuable for reliable online enforcement decisions, while the latter is valuable for deep offline analysis, continuous validation, and incident response.

**Comment by bureado** [comment]
Related to https://github.com/slsa-framework/slsa/issues/101.

See https://github.com/kpcyrd/rebuilderd and https://blog.sigstore.dev/celebrating-1-000-000-entries-in-rekor-1950b7c150df?gi=2c451ac76d4a

I might not be aware of everything that's happening in this space, but overall while I don't think we're too far from such a reality, right now  I don't think individuals should be expected to push provenance information and other attestations when they push an artifact. I think that's the sentiment in this and other threads as well. I can think of two groups of people to work with. One is build system operators, thinking OBS, Koji, Debian build system, etc. - likely QA folks and/or people involved in efforts such as Reproducible Builds. Can the tooling and processes be extended to submit attestations to e.g., a rekor? Ideally in a way that facilitates client-side verification; in other words, imagine an sbuild/pbuilder process can submit an entire .changes file to such a store in a way that an apt process client-side can be trivially extended to verify on install? See https://github.com/kpcyrd/pacman-bintrans.

The other group of people that could be interesting here are aggregators (there has to be a better term for this, my apologies) - thinking of Tidelift, Artifact Hub, Repology, etc. - people that "see" the packages coming through and that have their own processes/definition of trust for those packages. Maybe _they_ can ship attestations to a store, and maybe you can have a qualifier for those attestations, for example org: foo has seen package: bar:1.0 with hash: xyz, built using: <provenance information>, and we have seen this through: (check at least one) a) independent rebuild, b) dedicated trust process, c) something else. Happy to chat more. 2e-2.

**Comment by marcofranssen** [comment]
Wrote 2 blogs on the topic that might help in further defining a standard to address this.

- https://marcofranssen.nl/secure-your-software-supply-chain-using-sigstore-and-github-actions
- https://marcofranssen.nl/oci-as-attestations-storage-for-your-packages

**Comment by mlieberman85** [comment]
I am going to bump this back as part of the 1.0 work. There has been some discussion about either providing best practices, or pointing to good examples of distributing SLSA provenance so that folks who are building tooling for package ecosystems and other artifact distribution can implement SLSA distribution and discovery.

**Comment by di** [comment]
I self-assigned this, I am drafting a "Distributing Artifacts" page that should fit nicely between [Producing Artifacts](https://slsa.dev/spec/v1.0/requirements) and [Verifying Artifacts](https://slsa.dev/spec/v1.0/verifying-artifacts).

**Comment by di** [comment]
PR is here: https://github.com/slsa-framework/slsa/pull/673

**Comment by di** [comment]
#673 is getting close to being merged, if anyone here would like a chance to review beforehand.
